File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.js
// 2트: 
function solution(park, routes) {

  let [height, width] = [park.length, park[0].length]

  function canMove(sx,sy,direct,amount) {
    ways = [] // 초기화

    for (let i=1; i<=amount; i++) {
      if (direct === 'S') { sx += 1 }
      if (direct === 'N') { sx -= 1 }
      if (direct === 'W') { sy -= 1 }
      if (direct === 'E') { sy += 1 }
      ways.push([sx,sy]) // [0,1]
    }
  
    for (let way of ways) {
      if (way[0] >= height || way[0] < 0 ) return false;
      if (way[1] >= width || way[1] < 0 ) return false;
      if (park[way[0]][way[1]] === 'X') return false;
    }
    return true;
  }

  // step1: 시작 좌표 찾기
  let now = [0,0]
  for (let i=0; i<height; i++) {
    if (park[i].indexOf('S')>=0) now[0] = i
    for (let j=0; j<width; j++) {
      if (park[i][j] === 'S') now[1] = j
    }
  }

  // step2: routes 순회
  let ways = []
  for (let route of routes) {
    let [dir, num] = route.split(" ")
    if (canMove(now[0],now[1],dir,num)) {
      now[0] = ways[ways.length-1][0]
      now[1] = ways[ways.length-1][1]
    }
  }

  return now
}

console.log(solution(["SOO","OOO","OOO"],["E 2","S 2","W 1"]))
console.log(solution(["SOO","OXX","OOO"],["E 2","S 2","W 1"]))
console.log(solution(["OSO","OOO","OXO","OOO"],["E 2","S 3","W 1"]))

File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
# 아래 조건에 하나라도 걸리면, 가다가 마는게 아니라 아예 그 명령 수행 x
# 즉, 해당 명령의 '경로'가 이상하지 않는지 확인해야 함
    # 주어진 방향으로 이동할 때 공원을 벗어나는지 확인
    # 주어진 방향으로 이동 중 장애물을 만나는지 확인

# 1. s의 좌표 찾기

def solution(park, routes):
    way_dict = {'N': [-1,0], 'S': [1,0], 'W': [0,-1], 'E': [0,1]}

    for i in range(len(park)):
        for j in range(len(park[0])):
            if park[i][j] == 'S':
                sx, sy = i, j

    for route in routes:
        tx, ty = sx, sy # 임시
        flag = True
        for k in range(int(route[-1])):
            tx += way_dict[route[0]][0]
            ty += way_dict[route[0]][1]
            if (tx<0 or tx>=len(park)):
                flag= False
                break
            if (ty<0 or ty>= len(park[0])):
                flag= False
                break
            if park[tx][ty] == 'X':
                flag= False
                break

        if flag:
            sx, sy = tx, ty # 그제서야 이동시킴

    return [sx,sy]

# print(solution(["SOO","OOO","OOO"],["E 2","S 2","W 1"]))
# print(solution(["SOO","OXX","OOO"],["E 2","S 2","W 1"]))
print(solution(["OSO","OOO","OXO","OOO"],["E 2","S 3","W 1"]))